File: sdk/julia/src/bare_metal.jl
# Bare Metal & Colocation: reinstall, diagnostics, rescue, server info, IPs, VLANs, power
struct BareMetalResource
    http::HttpClient
end

# GET /api/service/{id}/reinstall/templates - List OS templates and recipes
function list_reinstall_templates(r::BareMetalResource; kwargs...)
    params = Dict(kwargs)
    http_get(r.http, "/api/service/{id}/reinstall/templates"; params = params)
end

# POST /api/service/{id}/diag/cancel - Cancel diagnostics operation
function cancel_diagnostics(r::BareMetalResource; kwargs...)
    params = Dict(kwargs)
    http_post(r.http, "/api/service/{id}/diag/cancel"; params = params)
end

# GET /api/service/{id}/diag/templates - List diagnostics templates
function list_diagnostics_templates(r::BareMetalResource; kwargs...)
    params = Dict(kwargs)
    http_get(r.http, "/api/service/{id}/diag/templates"; params = params)
end

# GET /api/service/{id}/diag - Show diagnostic operation status
function get_diagnostics_status(r::BareMetalResource; kwargs...)
    params = Dict(kwargs)
    http_get(r.http, "/api/service/{id}/diag"; params = params)
end

# POST /api/service/{id}/diag - Start new diagnostic operation
function run_diagnostics(r::BareMetalResource; kwargs...)
    params = Dict(kwargs)
    http_post(r.http, "/api/service/{id}/diag"; params = params)
end

# GET /api/service/{id}/rescue/templates - List rescue templates
function list_rescue_templates(r::BareMetalResource; kwargs...)
    params = Dict(kwargs)
    http_get(r.http, "/api/service/{id}/rescue/templates"; params = params)
end

# GET /api/service/{id}/rescue - Show rescue operation status
function get_rescue_status(r::BareMetalResource; kwargs...)
    params = Dict(kwargs)
    http_get(r.http, "/api/service/{id}/rescue"; params = params)
end

# POST /api/service/{id}/rescue - Start rescue operation
function rescue(r::BareMetalResource; kwargs...)
    params = Dict(kwargs)
    http_post(r.http, "/api/service/{id}/rescue"; params = params)
end

# POST /api/service/{id}/rescue/cancel - Cancel rescue operation
function cancel_rescue(r::BareMetalResource; kwargs...)
    params = Dict(kwargs)
    http_post(r.http, "/api/service/{id}/rescue/cancel"; params = params)
end

# GET /api/service/{id}/info - Get server hardware and OS details
function get_server_info(r::BareMetalResource; kwargs...)
    params = Dict(kwargs)
    http_get(r.http, "/api/service/{id}/info"; params = params)
end

# POST /api/service/{id}/hostname - Update server hostname
function update_hostname(r::BareMetalResource; kwargs...)
    params = Dict(kwargs)
    http_post(r.http, "/api/service/{id}/hostname"; params = params)
end

# GET /api/service/{id}/ips - List server IPs
function list_ips(r::BareMetalResource; kwargs...)
    params = Dict(kwargs)
    http_get(r.http, "/api/service/{id}/ips"; params = params)
end

# POST /api/service/{id}/ips - Add new IP to server
function add_ip(r::BareMetalResource; kwargs...)
    params = Dict(kwargs)
    http_post(r.http, "/api/service/{id}/ips"; params = params)
end

# GET /api/service/{id}/vlans - List available VLANs
function list_vlans(r::BareMetalResource; kwargs...)
    params = Dict(kwargs)
    http_get(r.http, "/api/service/{id}/vlans"; params = params)
end

# GET /api/service/{id}/ips/{ip} - Get IP details
function get_ip_details(r::BareMetalResource; kwargs...)
    params = Dict(kwargs)
    http_get(r.http, "/api/service/{id}/ips/{ip}"; params = params)
end

# PUT /api/service/{id}/ips/{ip} - Edit IP
function edit_ip(r::BareMetalResource; kwargs...)
    params = Dict(kwargs)
    http_put(r.http, "/api/service/{id}/ips/{ip}"; params = params)
end

# GET /api/service/{id}/reinstall - Return installation options
function get_reinstall_details(r::BareMetalResource; kwargs...)
    params = Dict(kwargs)
    http_get(r.http, "/api/service/{id}/reinstall"; params = params)
end

# POST /api/service/{id}/reinstall - Reinstall server
function reinstall(r::BareMetalResource; kwargs...)
    params = Dict(kwargs)
    http_post(r.http, "/api/service/{id}/reinstall"; params = params)
end

# GET /api/service/{id}/status - Return server power status
function get_status(r::BareMetalResource; kwargs...)
    params = Dict(kwargs)
    http_get(r.http, "/api/service/{id}/status"; params = params)
end

# POST /api/service/{id}/power/{action} - Power action (start/stop/reboot/reset/shutdown)
function power_action(r::BareMetalResource; kwargs...)
    params = Dict(kwargs)
    http_post(r.http, "/api/service/{id}/power/{action}"; params = params)
end

# GET /api/serverstock - List servers in stock
function list_servers_in_stock(r::BareMetalResource; kwargs...)
    params = Dict(kwargs)
    http_get(r.http, "/api/serverstock"; params = params)
end

File: sdk/dart/lib/src/bare_metal.dart
import 'http_client.dart';

/// Bare Metal & Colocation: reinstall, diagnostics, rescue, server info, IPs, VLANs, power
class BareMetalResource {
  final HttpClient httpClient;
  BareMetalResource(this.httpClient);

  /// GET /api/service/{id}/reinstall/templates - List OS templates and recipes
  Future<Map<String, dynamic>> listReinstallTemplates([Map<String, dynamic>? params]) async {
    return httpClient.get('/api/service/{id}/reinstall/templates', params);
  }

  /// POST /api/service/{id}/diag/cancel - Cancel diagnostics operation
  Future<Map<String, dynamic>> cancelDiagnostics([Map<String, dynamic>? params]) async {
    return httpClient.post('/api/service/{id}/diag/cancel', params);
  }

  /// GET /api/service/{id}/diag/templates - List diagnostics templates
  Future<Map<String, dynamic>> listDiagnosticsTemplates([Map<String, dynamic>? params]) async {
    return httpClient.get('/api/service/{id}/diag/templates', params);
  }

  /// GET /api/service/{id}/diag - Show diagnostic operation status
  Future<Map<String, dynamic>> getDiagnosticsStatus([Map<String, dynamic>? params]) async {
    return httpClient.get('/api/service/{id}/diag', params);
  }

  /// POST /api/service/{id}/diag - Start new diagnostic operation
  Future<Map<String, dynamic>> runDiagnostics([Map<String, dynamic>? params]) async {
    return httpClient.post('/api/service/{id}/diag', params);
  }

  /// GET /api/service/{id}/rescue/templates - List rescue templates
  Future<Map<String, dynamic>> listRescueTemplates([Map<String, dynamic>? params]) async {
    return httpClient.get('/api/service/{id}/rescue/templates', params);
  }

  /// GET /api/service/{id}/rescue - Show rescue operation status
  Future<Map<String, dynamic>> getRescueStatus([Map<String, dynamic>? params]) async {
    return httpClient.get('/api/service/{id}/rescue', params);
  }

  /// POST /api/service/{id}/rescue - Start rescue operation
  Future<Map<String, dynamic>> rescue([Map<String, dynamic>? params]) async {
    return httpClient.post('/api/service/{id}/rescue', params);
  }

  /// POST /api/service/{id}/rescue/cancel - Cancel rescue operation
  Future<Map<String, dynamic>> cancelRescue([Map<String, dynamic>? params]) async {
    return httpClient.post('/api/service/{id}/rescue/cancel', params);
  }

  /// GET /api/service/{id}/info - Get server hardware and OS details
  Future<Map<String, dynamic>> getServerInfo([Map<String, dynamic>? params]) async {
    return httpClient.get('/api/service/{id}/info', params);
  }

  /// POST /api/service/{id}/hostname - Update server hostname
  Future<Map<String, dynamic>> updateHostname([Map<String, dynamic>? params]) async {
    return httpClient.post('/api/service/{id}/hostname', params);
  }

  /// GET /api/service/{id}/ips - List server IPs
  Future<Map<String, dynamic>> listIps([Map<String, dynamic>? params]) async {
    return httpClient.get('/api/service/{id}/ips', params);
  }

  /// POST /api/service/{id}/ips - Add new IP to server
  Future<Map<String, dynamic>> addIp([Map<String, dynamic>? params]) async {
    return httpClient.post('/api/service/{id}/ips', params);
  }

  /// GET /api/service/{id}/vlans - List available VLANs
  Future<Map<String, dynamic>> listVlans([Map<String, dynamic>? params]) async {
    return httpClient.get('/api/service/{id}/vlans', params);
  }

  /// GET /api/service/{id}/ips/{ip} - Get IP details
  Future<Map<String, dynamic>> getIpDetails([Map<String, dynamic>? params]) async {
    return httpClient.get('/api/service/{id}/ips/{ip}', params);
  }

  /// PUT /api/service/{id}/ips/{ip} - Edit IP
  Future<Map<String, dynamic>> editIp([Map<String, dynamic>? params]) async {
    return httpClient.put('/api/service/{id}/ips/{ip}', params);
  }

  /// GET /api/service/{id}/reinstall - Return installation options
  Future<Map<String, dynamic>> getReinstallDetails([Map<String, dynamic>? params]) async {
    return httpClient.get('/api/service/{id}/reinstall', params);
  }

  /// POST /api/service/{id}/reinstall - Reinstall server
  Future<Map<String, dynamic>> reinstall([Map<String, dynamic>? params]) async {
    return httpClient.post('/api/service/{id}/reinstall', params);
  }

  /// GET /api/service/{id}/status - Return server power status
  Future<Map<String, dynamic>> getStatus([Map<String, dynamic>? params]) async {
    return httpClient.get('/api/service/{id}/status', params);
  }

  /// POST /api/service/{id}/power/{action} - Power action (start/stop/reboot/reset/shutdown)
  Future<Map<String, dynamic>> powerAction([Map<String, dynamic>? params]) async {
    return httpClient.post('/api/service/{id}/power/{action}', params);
  }

  /// GET /api/serverstock - List servers in stock
  Future<Map<String, dynamic>> listServersInStock([Map<String, dynamic>? params]) async {
    return httpClient.get('/api/serverstock', params);
  }

}
